thavalon-server/src/game/interactions.rs

```diff
@@ -0,0 +1,176 @@
+use futures::{future, TryFutureExt};
+use tokio::stream::{StreamExt, StreamMap};
+use tokio::sync::mpsc;
+
+use super::messages::{Action, GameError, Message};
+use super::PlayerId;
+
+#[async_trait]
+pub trait Interactions {
+    /// Send a message to a specific player
+    async fn send_to(&mut self, player: PlayerId, message: Message) -> Result<(), GameError>;
+
+    /// Send a message to all players
+    async fn send(&mut self, message: Message) -> Result<(), GameError>;
+
+    /// Receives messages from players until one satisfies the given function. The function can either return `Ok`, in which case the
+    /// result is immediately returned, or `Err`, in which case the error message is sent to the player and this waits for the next message.
+    ///
+    /// Think of this like a one-off `filter_map` operation.
+    async fn receive<F, R>(&mut self, mut f: F) -> Result<R, GameError>
+    where
+        R: Send,
+        F: FnMut(PlayerId, Action) -> Result<R, String> + Send;
+}
+
+/// An Interactions that uses per-player MPSC channels
+pub struct ChannelInteractions {
+    inbox: StreamMap<PlayerId, mpsc::Receiver<Action>>,
+    outbox: HashMap<PlayerId, mpsc::Sender<Message>>,
+}
+
+impl ChannelInteractions {
+    pub fn new() -> ChannelInteractions {
+        ChannelInteractions {
+            inbox: StreamMap::new(),
+            outbox: HashMap::new(),
+        }
+    }
+
+    pub fn add_player(
+        &mut self,
+        id: PlayerId,
+        incoming: mpsc::Receiver<Action>,
+        outgoing: mpsc::Sender<Message>,
+    ) {
+        self.inbox.insert(id, incoming);
+        self.outbox.insert(id, outgoing);
+    }
+}
+
+#[async_trait]
+impl Interactions for ChannelInteractions {
+    async fn send_to(&mut self, player: PlayerId, message: Message) -> Result<(), GameError> {
+        self.outbox
+            .get_mut(&player)
+            .unwrap()
+            .send(message)
+            .await
+            .map_err(|_| GameError::PlayerUnavailable { id: player })
+    }
+
+    async fn send(&mut self, message: Message) -> Result<(), GameError> {
+        let sends = self.outbox.iter_mut().map(|(id, sender)| {
+            sender
+                .send(message.clone())
+                .map_err(move |_| GameError::PlayerUnavailable { id: *id })
+        });
+        future::join_all(sends).await.into_iter().collect()
+    }
+
+    async fn receive<F, R>(&mut self, mut f: F) -> Result<R, GameError>
+    where
+        R: Send,
+        F: FnMut(PlayerId, Action) -> Result<R, String> + Send,
+    {
+        loop {
+            match self.inbox.next().await {
+                Some((player, action)) => match f(player, action) {
+                    Ok(result) => return Ok(result),
+                    Err(msg) => self.send_to(player, Message::Error(msg)).await?,
+                },
+                None => return Err(GameError::AllDisconnected),
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+pub(super) mod test {
```

Comment:
What is the pub(super) here?

Reply:
That makes it visible to the `game` module and sub-modules, but not the rest of the crate